src/tactic/omega/nat/main.lean

```diff
@@ -0,0 +1,175 @@
+import tactic.omega.nat.dnf
+import tactic.omega.nat.neg_elim
+
+open tactic 
+
+namespace omega
+namespace nat
+
+local notation `&` k    := preterm.cst k
+local infix `**`  : 300 := preterm.var 
+local notation t `+*` s := preterm.add t s
+local notation t `-*` s := preterm.sub t s
+
+local notation x `=*` y := form.eq x y
+local notation x `≤*` y := form.le x y
+local notation `¬*` p   := form.not p
+local notation p `∨*` q := form.or p q
+local notation p `∧*` q := form.and p q
+
+   
+run_cmd mk_simp_attr `sugar_nat 
+attribute [sugar_nat] 
+  not_le not_lt
+  nat.lt_iff_add_one_le
+  nat.succ_eq_add_one
+  or_false false_or
+  and_true true_and
+  ge gt mul_add add_mul mul_comm
+  classical.imp_iff_not_or
+  classical.iff_iff_not_or_and_or_not
+
+meta def desugar := `[try {simp only with sugar_nat}]
+
+lemma univ_close_of_unsat_neg_elim_not (m) (p : form) :
+  (neg_elim (¬* p)).unsat → univ_close p (λ _, 0) m :=  
+begin
+  intro h1, apply univ_close_of_valid,
+  apply valid_of_unsat_not, intro h2, apply h1,
+  apply form.sat_of_implies_of_sat implies_neg_elim h2,
+end
+
+meta def preterm.prove_sub_free : preterm → tactic expr 
+| (& m)    := return `(trivial)
+| (m ** n) := return `(trivial)
+| (t +* s) := 
+  do x ← preterm.prove_sub_free t,
+     y ← preterm.prove_sub_free s,
+     return `(@and.intro (preterm.sub_free %%`(t)) 
+       (preterm.sub_free %%`(s)) %%x %%y) 
+| (_ -* _) := failed
+
+meta def prove_neg_free : form → tactic expr 
+| (t =* s) := return `(trivial)
+| (t ≤* s) := return `(trivial)
+| (p ∨* q) := 
+  do x ← prove_neg_free p,
+     y ← prove_neg_free q,
+     return `(@and.intro (form.neg_free %%`(p)) 
+       (form.neg_free %%`(q)) %%x %%y) 
+| (p ∧* q) := 
+  do x ← prove_neg_free p,
+     y ← prove_neg_free q,
+     return `(@and.intro (form.neg_free %%`(p)) 
+       (form.neg_free %%`(q)) %%x %%y) 
+| _        := failed
+
+meta def prove_sub_free : form → tactic expr 
+| (t =* s) := 
+  do x ← preterm.prove_sub_free t,
+     y ← preterm.prove_sub_free s,
+     return `(@and.intro (preterm.sub_free %%`(t)) 
+       (preterm.sub_free %%`(s)) %%x %%y) 
+| (t ≤* s) := 
+  do x ← preterm.prove_sub_free t,
+     y ← preterm.prove_sub_free s,
+     return `(@and.intro (preterm.sub_free %%`(t)) 
+       (preterm.sub_free %%`(s)) %%x %%y) 
+| (¬*p) := prove_sub_free p
+| (p ∨* q) := 
+  do x ← prove_sub_free p,
+     y ← prove_sub_free q,
+     return `(@and.intro (form.sub_free %%`(p)) 
+       (form.sub_free %%`(q)) %%x %%y) 
+| (p ∧* q) := 
+  do x ← prove_sub_free p,
+     y ← prove_sub_free q,
+     return `(@and.intro (form.sub_free %%`(p)) 
+       (form.sub_free %%`(q)) %%x %%y) 
+
+/- Given a p : form, return the expr of a term t : p.unsat,
+   where p is subtraction- and negation-free. -/
+meta def prove_unsat_sub_free (p : form) : tactic expr :=  
+do x ← prove_neg_free p,
+   y ← prove_sub_free p,
+   z ← prove_unsats (dnf p),
+   return `(unsat_of_unsat_dnf %%`(p) %%x %%y %%z)
```

Comment:
This seems suspicious. You are putting `x` in the final proof term, a proof that `p` is `neg_free`, but you should know at this point that it must be neg_free by construction, and hence a proof is not needed. That is, this gets called from `univ_close_of_unsat_neg_elim_not `, which instantiates this with `neg_elim (¬* p)`, which has been proven `neg_free` in general. You may have to compose the theorems `univ_close_of_unsat_neg_elim_not`, `unsat_of_unsat_sub_elim`, `unsat_of_unsat_dnf` into one lemma to make this manifest though.

Reply:
Yes, in principle we know that `prove_unsat_sub_free` will always be called with a negation-free argument, so there should be no need to construct an additional proof that it is free of negations. The complication here is that this 'composed' theorem cannot be stated (at least not without something like an additional `nat` argument to indicate the number of subtraction eliminations) because we do not know a priori how many times `prove_unsat_neg_free` will call itself (and hence, how many times `unsat_of_unsat_sub_elim` will be applied).

One alternative is to change the definition of `sub_elim` from the current single-step elimination to a recursive function that eliminates all subtractions once and for all. This runs into termination proof issues for which I have not yet found an easy solution.

Another alternative is to iteratively build and pass down the proof that the argument is negation-free with each recursive call. I'm not sure this would be an improvement over the current version, though.